### PR TITLE
Generate numeric body and MAC

### PR DESCRIPTION
**Pull request by alexusmmunn**
## Contribution Guideline Acceptance

- [x] I have read and understand the [Individual Contributors License Agreement](CONTRIBUTING.md).
- [x] I accept the [Individual Contributors License Agreement](CONTRIBUTING.md) on the entire contents of this pull request.

## Proposed changes

Add functionality to nexus-python to generate numeric body and MAC address for passthrough keycodes

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have added unit tests with >= 90% coverage for all files modified or added by these changes
- [x] I have designed my tests to prove that my fix is effective or that my feature works as described
- [x] I have added docstrings to all new functions describing their purpose, parameters, and return value
- [ ] I have updated or added supplemental documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Further documentation can be found [here](https://github.com/angaza/payg-backend/pull/10643). This PR addresses goal 2.

### Reviewers:

This will be filled by the people who review this particular Pull Request.

### Attribution

This PR template derived from:
https://github.com/skcript/PR-Template

Original template is:

Copyright (c) 2017 Skcript

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

### CODE REVIEW

File: nexus_keycode/protocols/passthrough_uart.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import siphash
+
+DEV_KEY = "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
```
Comment (jjmilburn) [inline nexus_keycode/protocols/passthrough_uart.py:3]: Can also do:

```
DEV_KEY = "\x00" * 16
```

File: nexus_keycode/protocols/passthrough_uart.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def compute_uart_security_key(secret_key):
+    """Use a given secret key to generate a UART security key
+    :param secret_key: secret key used to generate UART security key
+    :type secret_key: str
+    """
+
+    # Split given key in half
+    key_part_a = secret_key[0 : len(secret_key) // 2]
```
Comment (jjmilburn) [inline nexus_keycode/protocols/passthrough_uart.py:13]: May want to add:

```
assert len(secret_key) == 16
```

File: nexus_keycode/test/protocols/test_passthrough_uart.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+from unittest import TestCase
+
+import nexus_keycode.protocols.passthrough_uart as uart
+
+
+class TestPassthroughUart(TestCase):
+    def test_passthrough_uart__expected_value_returned(self):
+        """ Test standard input into passthrough_uart and check
+            output"""
+        input_bytes = "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f"
```
Comment (jjmilburn) [inline nexus_keycode/test/protocols/test_passthrough_uart.py:10]: For consistency with existing test cases, I'd suggest using the following keys:

```
// 'input_bytes' is actually the 'base key' we are deriving the uart security key from
input_bytes = "\x01" * 14 + "\x43\x51"
```

The expected result is then:

```
uart_security_key = "\x12\xe4\x87\x62\x5c\x6b\x88\xf4\x1e\xe4\x0b\x16\xb4\xc9\x84\xf2"
```


from https://github.com/angaza/payg-embedded/blob/master/projects/Angaza_Portable_Library/src/angaza_test/test_payg_uart_pro.c#L49

**Comment by jjmilburn** [comment]
Dropped a few comments - also noting that what's being generated here is a ["Message Authentication Code"](https://en.wikipedia.org/wiki/Message_authentication_code), not a "MAC Address" - suggest changing the title!

File: nexus_keycode/protocols/utils.py [hunk 4]
```diff
@@ -58,3 +58,14 @@ def chunk(iteration):
...
 
     return output_bits[:output_len]
+
+
+def generate_mac(input_val, secret_key):
+    """
+    :type input_val: 'str'
+    :type secret_key: 'str'
+    """
+    key_bytes = secret_key.encode("utf-8")
```
Comment (jjmilburn) [inline nexus_keycode/protocols/utils.py:68]: This likely *works*, but the neither the secret key nor the input data should be `utf-8`. We expect to provide a secret key that is a raw, 16-byte value (e.g. `b"\x01\x02\x03... etc`).

By encoding it as `utf-8`, you are telling Python 'treat these bytes as representing a utf-8 string', which means 'when you see bytes in the string that represent UTF-8 codepoints, display the relevant character'. Heres an example:

```
>>> key = "\x01\x02\x03\x21\xc2\xb0"
>>> # \x01\x02\x03 don't have special meaning in UTF-8. \x21 is "!" in both ASCII and UTF-8. '\xc2\xb0' is "degree sign" in UTF-8
...
>>> encoded_key = key.encode('utf-8')
>>> encoded_key
b'\x01\x02\x03!\xc3\x82\xc2\xb0'
>>> encoded_key.decode('utf-8')
'\x01\x02\x03!Â°'
>>> # notice that python now shows "Â°, thats unexpected
... 
>>> decoded_key = encoded_key.decode('utf-8')
>>> str(decoded_key)
'\x01\x02\x03!Â°'
```

In other words, you *don't* want an encoding, you want raw bytes here. These bytes don't represent UTF-8 coded text.

Comment (jjmilburn) [reply nexus_keycode/protocols/utils.py:68]: Concrete suggestion, though:

```
def generate_mac(input_val, secret_key):
    """
    :type input_val: 'bytes'
    :type secret_key: 'bytes'
    """
    function = siphash.SipHash_2_4(key_bytes, int_to_bytes(input_bytes))
    return u"{:06d}".format(function.hash() & 0xFFFFFFFF)[-6:]
```

File: nexus_keycode/protocols/passthrough_uart.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import siphash
+
+NEXUS_INTEGRITY_CHECK_FIXED_00_KEY = "\x00" * 16
```
Comment (jjmilburn) [inline nexus_keycode/protocols/passthrough_uart.py:3]: To make both Python2 and Python3 happy here, you can make this change:

```
NEXUS_INTEGRITY_CHECK_FIXED_00_KEY = b"\x00" * 16
```

(notice the 'b' in front, indicating 'bytestring')

File: nexus_keycode/protocols/passthrough_uart.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+
+def compute_uart_security_key(secret_key):
+    """Use a given secret key to generate a UART security key
+    :param secret_key: secret key used to generate UART security key
+    :type secret_key: str
+    """
+
+    # Only 16 bit keys are accepted
```
Comment (jjmilburn) [inline nexus_keycode/protocols/passthrough_uart.py:12]: Nit: 16 **byte** keys

File: nexus_keycode/test/protocols/test_passthrough_uart.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class TestPassthroughUart(TestCase):
+    def test_passthrough_uart__expected_value_returned(self):
+        """ Test standard input into passthrough_uart and check
+            output"""
+        # 'input_bytes' is actually the 'base key' we are deriving
+        # the uart security key from
+        input_bytes = "\x01" * 14 + "\x43\x51"
+        uart_security_key = (
```
Comment (jjmilburn) [inline nexus_keycode/test/protocols/test_passthrough_uart.py:13]: Similar to comments above - Python 3 expects bytestrings. You can tell Python 3 you are dealing with bytestrings by prepending `b` to the string, e.g.

```
        input_bytes = b"\x01" * 14 + b"\x43\x51"
        uart_security_key = (
            b"\x12\xe4\x87\x62\x5c\x6b\x88\xf4\x1e\xe4\x0b\x16\xb4\xc9\x84\xf2"
        )
```
image

https://blog.feabhas.com/2019/02/python-3-unicode-and-byte-strings/

File: nexus_keycode/test/protocols/test_utils.py [hunk 8]
```diff
@@ -22,3 +22,10 @@ def test_pseudorandom_bits__varied_seeds__output_bits_are_expected(self):
             output = pseudorandom_bits(seed, expected.len)
 
             self.assertEqual(output, expected)
+
+    def test_generate_mac__standard_input__output_expected(self):
+        input_val = "\x00"
```
Comment (jjmilburn) [inline nexus_keycode/test/protocols/test_utils.py:27]: [Same comment](https://github.com/angaza/nexus-python/pull/12/files#r676227116) applies here

File: nexus_keycode/protocols/utils.py (hunk 4 above)
Comment (jjmilburn) [reply nexus_keycode/protocols/utils.py:68]: Note: This set of comments is optional, the code you have now works, but if we know we aren't dealing with a UTF-8 string at any point along the way, and the data is always bytestrings, it seems clearer to just skip the `encode` step (which *will* attempt to turn the data into a bytestring), and make sure the data being handled by `generate_mac` is a bytestring to begin with (rather than trying to convert it inside the function).

File: nexus_keycode/protocols/utils.py [hunk 9]
```diff
@@ -58,3 +58,13 @@ def chunk(iteration):
...
+
+
+def generate_mac(input_val, secret_key):
+    """
+    :type input_val: 'byte'
+    :type secret_key: 'byte'
+    """
+    # Return the last 6 digits of the siphash result
+    function = siphash.SipHash_2_4(secret_key, input_val)
+    return u"{:06d}".format(function.hash() & 0xFFFFFFFF)[-6:]
```
Comment (jjmilburn) [inline nexus_keycode/protocols/utils.py:70]: masks lower 32 bits, and returns 6 digits :+1:

Comment (jjmilburn) [reply nexus_keycode/protocols/utils.py:70]: So, this is actually a `full` protocol MAC, it is not the same MAC we use for all protocols here (for example, `small` protocol does not use this MAC).

This makes me wonder whether this belongs in the `utils` folder, but I think its fine since we need it for a separate use case here. Can we name it something a bit more descriptive, e.g. "generate_base10_mac" or something?

**Comment by jjmilburn** [comment]
Minor comment on making `generate_mac` less ambiguous (someone looking in the utils folder might think `generate_mac` is used for the protocols here, but its not). +1!
